libraries/lib-wave-track/CachingPlayableSequence.cpp

```diff
@@ -0,0 +1,115 @@
+#include "CachingPlayableSequence.h"
+#include "AudioSegmentSampleView.h"
+#include "WaveTrack.h"
+
+#include <cassert>
+
+namespace
+{
+void FillBufferFromTrackBlockSequence(
+   const TrackSampleView& track, float* buffer, size_t bufferSize)
+{
+   size_t written = 0;
+   for (const auto& segment : track)
+   {
+      size_t toWrite = std::min(
+         segment.GetSampleCount().as_long_long(),
+         static_cast<long long>(bufferSize - written));
```

Comment:
Again, use `limitSampleBufferSize`

Reply:
You assume `bufferSize >= written` but what proves that?  Should there be a check?

Reply:
This is based on the assumption that `AudioSegmentSampleView::Copy` does what it says, namely not returning more than `toWrite` samples. A unit-test proves this, which I'd argue is enough ?

Reply:
Tests are just one line of defense ensuring code quality and raising our confidence.  I think they should be complemented with assertions in the program.  (I mean both `assert` and also the informal statement of pres and posts and invariants.)

When I write assertions it means I'm very serious about persuading the unfamiliar reader (who might be you, or future me) that the code is CORRECT, because I was working hard at the time to convince myself first and to make more explicit the assumptions that I relied on.

A test, though automated, is the weaker form of "well it seems to work".

True, human reasoning is still fallible and tests may still catch mistakes (perhaps because of future modifications by someone understanding the code imprefectly).

Still -- write contracts first, I think, and a well thought contract is then a guide for writing the test cases themselves.

Some correspondence between test cases and formal or informal assertions in the program is a good idea.

Reply:
Numerous times, I have seen well-thought assertions fail on the first run.

Reply:
Yes, we are fallible, but still we should value well thought assertions and make attempts at rigor.

As I said, I think contracts as a guide to what cases should be tested.  Or if you like the TDD approach of writing failing tests first, and then then changing the program to succeed, I think then you should say something in the contract too.

Tests might be "executable documentation" but I like documentation in the header itself, explaining all that is important in the interface in one place.

Reply:
I actually found an alternative implementation, where it is obvious that `written < bufferSize`, and which does not unnecessarily iterate over segments after if `written` already `== bufferSize`.

But I'd have complied.